### ISSUE TITLE
Axioms admit more transitions than intended

### ISSUE THREAD

**Issue by ttuegel** [kframework/kore#672]
The encoding of K rules into Kore axioms admits more transitions than intended.

#### Notation

`∅` is the null `Set`
`{ x₁, x₂, ... }` is the `Set` of elements `x₁`, `x₂`, ...
`_∪_` is the usual `Set` union symbol
`_∈_` is the `Bool`-valued `Set` membership predicate

### Problem

Consider the axiom
```
(A).    ∀ x S.  { x } ∪ S  →∙  φ(S)
```
where `φ(S)` is an arbitrary pattern in which `S` occurs free. This axiom is would be generated by K to represent a transition
```
(T1).    { a }  →∙  φ(∅)
```
which we can see by instantiating axiom (A) with `x = a  ∧  S = ∅`. However, we can also instantiate (A) with `x = a  ∧  S = { a }` to obtain the transition
```
(T2).    { a }  →∙  φ({ a })
```
because `_∪_` is idempotent. Although the user intended only transition (T1), the transition system described by (A) also includes (T2).

This problem is not unique to `Set`, but includes `Map` and probably any other domain with an idempotent, associative, commutative operator such as `_∪_`.

### Solution

From the backend's perspective, we must produce the correct unifier,
```
(U1).    ⌈ { x } ∪ S  ∧  { a } ⌉  =  x = a  ∧  (S = ∅  ∨  S = { a })
```
so that we will simplify other expressions correctly, such as `{ x } ∪ S  =  { a }`. This is not satisfactory to the user, so we suggest adding an appropriate pre-condition to (A) to take the desired transition:
```
(A1).    ∀ x S.  (x ∉ S = true  ∧  { x } ∪ S) →∙ φ(S)  -- admits only (T1)
(A2).    ∀ x S.  (x ∈ S = true  ∧  { x } ∪ S) →∙ φ(S)  -- admits only (T2)
```
Unfortunately, this may require some semantics to be modified in order to be compatible with 
the Haskell backend. However, most rules of the form of (A) actually do _not_ have `S` occurring free in `φ`, so the scope of changes required may be minimal. We also note that, because the Haskell backend supports non-deterministic unification on `Set` and `Map`, rules which unify a `Set` or `Map` (instead of directly testing membership or lookup) may introduce undesirable redundant branches.

**Comment by virgil-serbanuta** [comment]
I'd have said that the semantics must be changed in order to be compatible with Matching Logic and the `_∪_` semantics, not with the Haskell backend, since the backend just implements ML... Or that the frontend must translate the user intent in proper ML/kore formulas, including the `x ∉ S` precondition, if that makes more sense.

**Comment by ttuegel** [comment]
@virgil-serbanuta I agree with your characterization (the user needs to express their intent more logically), but as a practical matter, this is only an issue for the Haskell backend because none of the other backends support non-deterministic unification.

**Comment by ttuegel** [comment]
This was fixed in: #782, #770, #758, #753, #711, and possibly others.
